x/evm/keeper/keeper.go

```diff
@@ -213,166 +174,295 @@ func (k Keeper) GetTxIndexTransient() uint64 {
-func (k Keeper) GetLogSizeTransient() uint64 {
-	store := k.Ctx().TransientStore(k.transientKey)
-	bz := store.Get(types.KeyPrefixTransientLogSize)
-	if len(bz) == 0 {
+// GetNonce returns the sequence number of an account, returns 0 if not exists.
+func (k *Keeper) GetNonce(ctx sdk.Context, addr common.Address) uint64 {
+	cosmosAddr := sdk.AccAddress(addr.Bytes())
+	acct := k.accountKeeper.GetAccount(ctx, cosmosAddr)
+	if acct == nil {
 		return 0
 	}
 
-	return sdk.BigEndianToUint64(bz)
+	ethAcct, ok := acct.(*ethermint.EthAccount)
+	if !ok {
+		return 0
+	}
+
+	return ethAcct.Sequence
 }
 
-// IncreaseLogSizeTransient fetches the current EVM log index from the transient store, increases its
-// value by one and then sets the new index back to the transient store.
-func (k Keeper) IncreaseLogSizeTransient() {
-	logSize := k.GetLogSizeTransient()
-	store := k.Ctx().TransientStore(k.transientKey)
-	store.Set(types.KeyPrefixTransientLogSize, sdk.Uint64ToBigEndian(logSize+1))
+// GetBalance load account's balance of gas token
+func (k *Keeper) GetBalance(ctx sdk.Context, addr common.Address) *big.Int {
+	cosmosAddr := sdk.AccAddress(addr.Bytes())
+	params := k.GetParams(ctx)
+	coin := k.bankKeeper.GetBalance(ctx, cosmosAddr, params.EvmDenom)
+	return coin.Amount.BigInt()
 }
 
 // ----------------------------------------------------------------------------
-// Storage
+// statedb.Keeper implementation
 // ----------------------------------------------------------------------------
 
-// GetAccountStorage return state storage associated with an account
-func (k Keeper) GetAccountStorage(ctx sdk.Context, address common.Address) (types.Storage, error) {
-	storage := types.Storage{}
+// GetAccount returns nil if account is not exist, returns error if it's not `EthAccount`
+func (k *Keeper) GetAccount(ctx sdk.Context, addr common.Address) (*statedb.Account, error) {
+	acct, err := k.GetAccountWithoutBalance(ctx, addr)
+	if acct == nil || err != nil {
+		return acct, err
+	}
 
-	err := k.ForEachStorage(address, func(key, value common.Hash) bool {
-		storage = append(storage, types.NewState(key, value))
-		return true
-	})
+	acct.Balance = k.GetBalance(ctx, addr)
+	return acct, nil
+}
+
+// GetAccountOrEmpty returns empty account if not exist, returns error if it's not `EthAccount`
+func (k *Keeper) GetAccountOrEmpty(ctx sdk.Context, addr common.Address) (statedb.Account, error) {
+	acct, err := k.GetAccount(ctx, addr)
 	if err != nil {
-		return types.Storage{}, err
+		return statedb.Account{}, err
 	}
-
-	return storage, nil
+	if acct == nil {
+		// empty account
+		return statedb.Account{
+			Balance:  new(big.Int),
+			CodeHash: emptyCodeHash,
+		}, nil
+	}
+	return *acct, nil
 }
 
-// ----------------------------------------------------------------------------
-// Account
-// ----------------------------------------------------------------------------
+func (k *Keeper) GetState(ctx sdk.Context, addr common.Address, key common.Hash) common.Hash {
```

Comment:
can you add comments on all public funcs here?

Reply:
done